src/openforms/analytics_tools/models.py

```diff
@@ -0,0 +1,312 @@
+        return (
+            self.matomo_url
+            and self.matomo_site_id
+            and self.enable_matomo_site_analytics
+            and self.analytics_cookie_consent_group
+        )
+
+    def piwik_can_be_enabled(self) -> bool:
+        return (
+            self.piwik_url
+            and self.piwik_site_id
+            and self.enable_piwik_site_analytics
+            and self.analytics_cookie_consent_group
+        )
+
+    def piwik_pro_can_be_enabled(self) -> bool:
+        return (
+            self.piwik_pro_url
+            and self.piwik_pro_site_id
+            and self.enable_piwik_pro_site_analytics
+            and self.analytics_cookie_consent_group
+        )
+
+    def siteimprove_can_be_enabled(self) -> bool:
+        return (
+            self.siteimprove_id
+            and self.enable_siteimprove_analytics
+            and self.analytics_cookie_consent_group
+        )
+
+    def google_analytics_can_be_enabled(self) -> bool:
+
+        return (
+            self.ga_code
+            and self.gtm_code
+            and self.enable_google_analytics
+            and self.analytics_cookie_consent_group
+        )
+
+    def update_analytical_cookies(self, cookies: dict, create: bool):
+        if create:
+            cookie_group = CookieGroup.objects.get(
+                id=self.analytics_cookie_consent_group.id
+            )
+            for cookie in cookies:
+                Cookie.objects.get_or_create(
+                    name=cookie["name"],
+                    path=cookie["path"],
+                    domain=settings.ALLOWED_HOSTS[0],
+                    cookiegroup=cookie_group,
+                )
+        else:
+            for cookie in cookies:
+                Cookie.objects.filter(name=cookie["name"]).delete()
+
+    def update_csp(self, csps: dict, create: bool):
+        if create:
+            for csp in csps:
+                CSPSetting.objects.get_or_create(
+                    directive=csp["directive"], value=csp["value"]
+                )
+        else:
+            for csp in csps:
+                CSPSetting.objects.filter(value=csp["value"]).delete()
+
+    def save(self, *args, **kwargs):
+        # If instance is being created, we can't find original values
+        if self._state.adding:
+            return super().save(*args, **kwargs)
+        original_object = self.__class__.objects.get(pk=self.pk)
+        # For each analytics provider, we check if :
+        # - the value of the provider's boolean field is changing
+        # - one of the fields required is not filled out while the providers's boolean field one is --> we need to reset the boolean field to False
+        if original_object.enable_google_analytics != self.enable_google_analytics or (
+            not self.google_analytics_can_be_enabled() and self.enable_google_analytics
+        ):
+            if not self.google_analytics_can_be_enabled():
+                self.enable_google_analytics = False
```

Comment:
this makes it possible for the end user to check a checkbox with incomplete configuration, and then the checkbox will be silently unchecked, right?

I think it would be better to implement the validation in the `def clean(...)` method and raise a `django.core.exceptions.ValidationError`, something along the lines "If you enable $provider, you must fill out all the required fields.". That provides explicit user feedback, and then in this `save` method, you only have to check the value of the `is_<provider>_enabled` and/or the boolean flag.

Reply:
Yes, I agree. It would also make the save method more readable